cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -0,0 +1,131 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <mutex>
+
+#include "arrow/api.h"
+#include "arrow/compute/api.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+struct UnionNode : ExecNode {
+  UnionNode(ExecNode* lhs_input, ExecNode* rhs_input, ExecContext* ctx)
+      : ExecNode(lhs_input->plan(), {lhs_input, rhs_input},
+                 {"left_input_union", "right_input_union"},
+                 /*output_schema=*/lhs_input->output_schema(),
+                 /*num_outputs=*/1),
+        ctx_(ctx) {}
+
+  const char* kind_name() override { return "UnionNode"; }
+
+  inline bool IsLeftInput(ExecNode* input) { return input == inputs_[0]; }
+
+  void InputReceived(ExecNode* input, int seq, ExecBatch batch) override {
+    ARROW_DCHECK(input == inputs_[0] || input == inputs_[1]);
+
+    if (finished_.is_finished()) {
+      return;
+    }
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      ++batch_count_;
+    }
+    outputs_[0]->InputReceived(this, seq, std::move(batch));
+  }
+
+  void ErrorReceived(ExecNode* input, Status error) override {
+    DCHECK_EQ(input, inputs_[0]);
+    outputs_[0]->ErrorReceived(this, std::move(error));
+
+    StopProducing();
+  }
+
+  void InputFinished(ExecNode* input, int num_total) override {
+    ARROW_DCHECK(input == inputs_[0] || input == inputs_[1]);
+    {
+      std::unique_lock<std::mutex> lk(mutex_);
+      ++input_count_;
+    }
+    std::unique_lock<std::mutex> lk(mutex_);
+    if (input_count_ == 2) {
+      finished_.MarkFinished();
+      outputs_[0]->InputFinished(this, batch_count_);
```

Comment:
InputFinished isn't guaranteed to be called after calls to InputReceived finish. Instead we should add `num_total` to a counter here, then when we've gotten InputFinished from all inputs, we can then call InputFinished on the output.

We shouldn't MarkFinished until we get all the batches. 

You can see this pattern in aggregate_node.cc: https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/exec/aggregate_node.cc#L483-L492 and https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/exec/aggregate_node.cc#L464-L475

Reply:
So it would look something like (assuming we use AtomicCounter)

```cpp
input_count_.SetTotal(2); // or inputs.size() if we want to make this a variadic node
// ...

// InputFinished
total_batches_ += num_total;
if (input_count_.Increment()) {
  outputs_[0]->InputFinished(this, total_batches_);
  if (batch_count.SetTotal(total_batches_)) {
    finished_.MarkFinished();
  }
}

// InputReceived
outputs_[0]->InputReceived(this, seq, std::move(batch));
if (batch_count_.Increment()) {
  finished_.MarkFinished();
}

// StopProducing
if (batch_count_.Cancel()) finished_.MarkFinished();
```

Reply:
I wonder if we could make some sort of testing node that purposely permuted the order of operations to try to catch this…

Reply:
Thanks I followed your suggest to update this PR